### PR TITLE
Fix Signature on Generic Expansion Methods (again)

### PR DESCRIPTION
**Pull request by jaredlll08**
This is the same PR as https://github.com/ZenCodeLang/ZenCode/pull/51, which was reverted due to failing tests, with the only change being:
```diff
	public Type getASMType() {
+		if(CompilerUtils.isPrimitive(elementType)) {
+			return context.getType(elementType);
+		}
		return Type.getType("[" + context.getDescriptor(elementType));
	}
```
in `ArrayHelperType.java`, all the tests pass.

### CODE REVIEW

File: ScriptingExample/src/test/java/org/openzen/zenscript/scriptingexample/tests/actual_test/arrays/ArrayOperatorTests.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+				.add("var mapper as function(index as usize, element as string) as string = (index as usize, element as string) as string => element + index + element;")
+				.add("var mapped = array.map<string>(mapper);")
+				.add("for element in mapped println(element);")
+				.execute(this);
+
+		logger.assertPrintOutputSize(3);
+		logger.assertPrintOutput(0, "A0A");
+		logger.assertPrintOutput(1, "B1B");
+		logger.assertPrintOutput(2, "C2C");
+	}
```
Comment (kindlich) [inline ScriptingExample/src/test/java/org/openzen/zenscript/scriptingexample/tests/actual_test/arrays/ArrayOperatorTests.java:122]: You could add tests for the other means of creating arrays (both single, and multi-dimensional):
- [x] Literal: `var x = [1, 2, 3]`
- [x] Constructor with default value
     - [x] 1D: `var x = new int[](size, defaultValue); println(x[0]);`
     - [x] 2D: `var x = new int[,](sizeA, sizeB, defaultValue); println(x[0,0]);`
- [x] Constructor with callback function `var x = new int[](size, (index as usize) => {return 10 as int;}); println(x[0]);`
- [ ] Constructor with projection function `var x = [1, 2, 3] as int[]; var y = new int[](x, (valueFromX as int) => {return 2 * valueFromX;}); println(y[0]);`

File: JavaBytecodeCompiler/src/main/java/org/openzen/zenscript/javabytecode/compiler/ArrayHelperType.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+		} else {
+			javaWriter.newArray(context.getType(elementType));
+		}
+	}
+
+	public Type getASMType() {
+		if(CompilerUtils.isPrimitive(elementType)) {
+			return context.getType(elementType);
+		}
+		return Type.getType("[" + context.getDescriptor(elementType));
```
Comment (kindlich) [inline JavaBytecodeCompiler/src/main/java/org/openzen/zenscript/javabytecode/compiler/ArrayHelperType.java:45]: I'm not too keen on the way this function is used atm.

Currently in one case it returns an array type and sometimes a primitive.
That looks to me as though back then I did screw up somewhere with the types and when these methods are called 🤔 


We could always return `context.getType(elementType)` here (and maybe rename it to getElementTypeAsASM or something).
Then, in ArrayInitializerHelper.java#195:
```diff
@@ -192,7 +192,7 @@ class ArrayInitializerHelper {
                } else {
                        visitMultiDimArray(javaWriter, sizeLocations, counterLocations, dim - 1, elementType, innermostFunction);
                }
-               javaWriter.arrayStore(elementType.getASMType());
+               javaWriter.arrayStore(currentArrayType.getASMType());
 
                //Return to the start
                javaWriter.iinc(forLoopCounter);
```

That way we would be consistent in always using the element type here..?

**Review by kindlich — APPROVED** [review]
Quick tests shows its working.

**Comment by jaredlll08** [comment]
This PR was inadvertently merged due to committing to the wrong branch, the push has been reverted and a new PR has been made to track this. 
https://github.com/ZenCodeLang/ZenCode/pull/123

**Pull request merged**